x-pack/plugins/ml/server/shared_services/shared_services.ts

```diff
@@ -18,40 +22,107 @@ import {
-  isMinimumLicense(): void;
-  getHasMlCapabilities(request: KibanaRequest): HasMlCapabilities;
+  getGuards(request: KibanaRequest): Guards;
 }
 
+interface OkParams {
+  scopedClient: IScopedClusterClient;
+}
+
+type OkCallback = (okParams: OkParams) => any;
+
 export function createSharedServices(
   mlLicense: MlServerLicense,
   spaces: SpacesPluginSetup | undefined,
   cloud: CloudSetup,
-  resolveMlCapabilities: ResolveMlCapabilities
+  resolveMlCapabilities: ResolveMlCapabilities,
+  getClusterClient: () => IClusterClient | null
 ): SharedServices {
+  const getRequestItems = getRequestItemsProvider(resolveMlCapabilities, getClusterClient);
   const { isFullLicense, isMinimumLicense } = licenseChecks(mlLicense);
-  const getHasMlCapabilities = hasMlCapabilitiesProvider(resolveMlCapabilities);
-  const checks: SharedServicesChecks = {
-    isFullLicense,
-    isMinimumLicense,
-    getHasMlCapabilities,
-  };
+
+  function getGuards(request: KibanaRequest): Guards {
+    const { hasMlCapabilities, scopedClient } = getRequestItems(request);
+    const asyncGuards: Array<Promise<void>> = [];
+
+    const guards: Guards = {
+      isMinimumLicense: () => {
+        isMinimumLicense();
+        return guards;
+      },
+      isFullLicense: () => {
+        isFullLicense();
+        return guards;
+      },
+      hasMlCapabilities: (caps: MlCapabilitiesKey[]) => {
+        asyncGuards.push(hasMlCapabilities(caps));
+        return guards;
+      },
+      async ok(callback: OkCallback) {
+        await Promise.all(asyncGuards);
+        return callback({ scopedClient });
+      },
+    };
+    return guards;
+  }
 
   return {
-    ...getJobServiceProvider(checks),
-    ...getAnomalyDetectorsProvider(checks),
-    ...getModulesProvider(checks),
-    ...getResultsServiceProvider(checks),
-    ...getMlSystemProvider(checks, mlLicense, spaces, cloud, resolveMlCapabilities),
+    ...getJobServiceProvider(getGuards),
+    ...getAnomalyDetectorsProvider(getGuards),
+    ...getModulesProvider(getGuards),
+    ...getResultsServiceProvider(getGuards),
+    ...getMlSystemProvider(getGuards, mlLicense, spaces, cloud, resolveMlCapabilities),
+  };
+}
+
+function getRequestItemsProvider(
+  resolveMlCapabilities: ResolveMlCapabilities,
+  getClusterClient: () => IClusterClient | null
+) {
+  return (request: KibanaRequest) => {
+    const getHasMlCapabilities = hasMlCapabilitiesProvider(resolveMlCapabilities);
+    let hasMlCapabilities: HasMlCapabilities;
+    let scopedClient: IScopedClusterClient;
+    // While https://github.com/elastic/kibana/issues/64588 exists we
+    // will not receive a real request object when being called from an alert.
+    // instead a dummy request object will be supplied
+    const clusterClient = getClusterClient();
+
+    if (clusterClient === null) {
+      throw new MLClusterClientUninitialized(`ML's cluster client has not been initialized`);
+    }
+
+    if (request instanceof KibanaRequest) {
```

Comment:
With this check, the request object is essentially optional. It would make sense to have the request param be optional in the consumer API functions for the providers.

Reply:
I'd prefer to keep the `request` object as required and have consumers supply a "fake" one to reiterate the fact that there is a problem with the permission checks when calling our functions from an alert.
The need for a fake request object should be temporary until this alerting issue has been addressed (https://github.com/elastic/kibana/issues/64588)